README.md
# Todo Tree

This extension quickly searches (using [ripgrep](https://github.com/BurntSushi/ripgrep)) your workspace for comment tags like TODO and FIXME, and displays them in a tree view in the activity bar. The view can be dragged out of the activity bar into the explorer pane (or anywhere else you would prefer it to be).

Clicking a TODO within the tree will open the file and put the cursor on the line containing the TODO.

Found TODOs can also be highlighted in open files.

*Please see the [wiki](https://github.com/Gruntfuggly/todo-tree/wiki/Configuration-Examples) for configuration examples.*

![screenshot](https://raw.githubusercontent.com/Gruntfuggly/todo-tree/master/resources/screenshot.png)

*Notes:*

- *User* `rg.conf` *files are ignored.*

## Highlighting

>*New!:* If you just want to set different colours for tags, you can now enable `todo-tree.highlights.useColourScheme`. This will apply a set of colours (which can be changed) to the tags in the order that they are defined.

Highlighting tags is configurable. Use `defaultHighlight` to set up highlights for all tags. If you need to configure individual tags differently, use `customHighlight`. If settings are not specified in `customHighlight`, the value from `defaultHighlight` is used.

Custom highlights can also be specified for sub tags (if used).

<sup>*Note: `defaultHighlight` is not applied to sub tags.*</sup>

Both `defaultHighlight` and `customHighlight` allow for the following settings:

`foreground` - used to set the foreground colour of the highlight in the editor and the marker in the ruler.

`background` - used to set the background colour of the highlight in the editor.

<sup>*Note: Foreground and background colours can be specified using [HTML/CSS colour names](https://en.wikipedia.org/wiki/Web_colors) (e.g. "Salmon"), RGB hex values (e.g. "#80FF00"), RGB CSS style values (e.g. "rgb(255,128,0)" or colours from the current theme, e.g. `peekViewResult.background`. See [Theme Color](https://code.visualstudio.com/api/references/theme-color) for the details. Hex and RGB values can also have an alpha specified, e.g. "#ff800080" or "rgba(255,128,0,0.5)".*</sup>

`opacity` - percentage value used with the background colour. 100% will produce an opaque background which will obscure selection and other decorations.

<sup>*Note: opacity can only be specified when hex or RGB colours are used.*</sup>

`fontWeight`, `fontStyle`, `textDecoration` - can be used to style the highlight with standard CSS values.

`borderRadius` - used to set the border radius of the background of the highlight.

`icon` - used to set a different icon in the tree view. Must be a valid octicon (see <https://octicons.github.com>) or codicon (see <https://microsoft.github.io/vscode-codicons/dist/codicon.html>). If using codicons, specify them in the format "$(*icon*)". The icon defaults to a tick if it's not valid. You can also use "todo-tree", or "todo-tree-filled" if you want to use the icon from the activity view.

`iconColour` - used to set the colour of the icon in the tree. If not specified, it will try to use the foreground colour or the background colour. Colour can be specified as per foreground and background colours, but see note below.

<sup>*Note: Theme colours are only supported when using codicons. Hex, RGB and HTML colours are only supported when using octicons or the default icon.*</sup>

`gutterIcon` - set to true to show the icon in the editor gutter.

<sup>*Note: Unfortunately, only octicons and the todo-tree icon can be displayed in the gutter.*</sup>

`rulerColour` - used to set the colour of the marker in the overview ruler. If not specified, it will default to use the foreground colour. Colour can be specified as per foreground and background colours.

`rulerOpacity` - used to set the opacity of the ruler markers.

<sup>*Note: Only works with hex and RGB colour settings.*</sup>

`rulerLane` - used to set the lane for the marker in the overview ruler. If not specified, it will default to the right hand lane. Use one of "left", "center", "right", or "full". You can also use "none" to disable the ruler markers.

`type` - used to control how much is highlighted in the editor. Valid values are:

- `tag` - highlights just the tag
- `text` - highlights the tag and any text after the tag
- `tag-and-comment` - highlights the comment characters (or the start of the match) and the tag
- `tag-and-subTag` - as above, but allows the sub tag to be highlight too (with colours defined in custom highlights)
- `text-and-comment` - highlights the comment characters (or the start of the match), the tag and the text after the tag
- `line` - highlights the entire line containing the tag
- `whole-line` - highlights the entire line containing the tag to the full width of the editor
- `capture-groups:n,m...` - highlights capture groups from the regex, where 'n' is the index into the regex
- `none` - disable highlightling in the document

`hideFromTree` - used to hide tags from the tree, but still highlight in files

`hideFromStatusBar` - prevents the tag from being included in the status bar counts

`hideFromActivityBar` - prevents the tag from being included in the activity bar badge count

Example:

```json
"todo-tree.highlights.defaultHighlight": {
    "icon": "alert",
    "type": "text",
    "foreground": "red",
    "background": "white",
    "opacity": 50,
    "iconColour": "blue"
},
"todo-tree.highlights.customHighlight": {
    "TODO": {
        "icon": "check",
        "type": "line"
    },
    "FIXME": {
        "foreground": "black",
        "iconColour": "yellow",
        "gutterIcon": true
    }
}
```

<sup>*Note: The highlight configuration is separate from the settings for the search. Adding settings in `customHighlight` does not automatically add the tags into `todo-tree.general.tags`.*</sup>

<sup>*Note: Using the `capture-groups` setting in `type` may have an impact on performance with large files.

## Installing

You can install the latest version of the extension via the Visual Studio Marketplace [here](https://marketplace.visualstudio.com/items?itemName=Gruntfuggly.todo-tree).

Alternatively, open Visual Studio code, press `Ctrl+P` or `Cmd+P` and type:

> ext install Gruntfuggly.todo-tree

### Source Code

The source code is available on GitHub [here](https://github.com/Gruntfuggly/todo-tree).

## Controls

The tree view header can contain the following buttons:

![collapse](https://raw.githubusercontent.com/Gruntfuggly/todo-tree/master/resources/button-icons/collapse.png) - Collapse all tree nodes</br>
![expand](https://raw.githubusercontent.com/Gruntfuggly/todo-tree/master/resources/button-icons/expand.png) - Expand all tree nodes</br>
![flat](https://raw.githubusercontent.com/Gruntfuggly/todo-tree/master/resources/button-icons/flat.png) - Show the tree view as a flat list, with the full filename for each TODO</br>
![tags](https://raw.githubusercontent.com/Gruntfuggly/todo-tree/master/resources/button-icons/tags.png) - Show the view as a list of tags</br>
![tree](https://raw.githubusercontent.com/Gruntfuggly/todo-tree/master/resources/button-icons/tree.png) - Show the tree view as a tree with expandable nodes for each folder (default)</br>
![tag](https://raw.githubusercontent.com/Gruntfuggly/todo-tree/master/resources/button-icons/tag.png) - Group the TODOs in the tree by the tag</br>
![notag](https://raw.githubusercontent.com/Gruntfuggly/todo-tree/master/resources/button-icons/notag.png) - Organise the TODOs by file (default)</br>
![filter](https://raw.githubusercontent.com/Gruntfuggly/todo-tree/master/resources/button-icons/filter.png) - Only show items in the tree which match the entered filter text</br>
![clear-filter](https://raw.githubusercontent.com/Gruntfuggly/todo-tree/master/resources/button-icons/clear-filter.png) - Remove any active filter</br>
![refresh](https://raw.githubusercontent.com/Gruntfuggly/todo-tree/master/resources/button-icons/refresh.png) - Rebuild the tree</br>
![scan-open-files](https://raw.githubusercontent.com/Gruntfuggly/todo-tree/master/resources/button-icons/scan-open-files.png) - Show tags from open files only</br>
![scan-current-file](https://raw.githubusercontent.com/Gruntfuggly/todo-tree/master/resources/button-icons/scan-current-file.png) - Show tags from the current file</br>
![scan-workspace](https://raw.githubusercontent.com/Gruntfuggly/todo-tree/master/resources/button-icons/scan-workspace-only.png) - Show tags from workspace only</br>
![scan-workspace](https://raw.githubusercontent.com/Gruntfuggly/todo-tree/master/resources/button-icons/scan-workspace.png) - Show tags from workspace and open files</br>
![reveal](https://raw.githubusercontent.com/Gruntfuggly/todo-tree/master/resources/button-icons/reveal.png) - Show the current file in the tree</br>
![export](https://raw.githubusercontent.com/Gruntfuggly/todo-tree/master/resources/button-icons/export.png) - Export the tree content to a file</br>

## Folder Filter Context Menu

Right clicking in the tree view will bring up a context menu with the following options:

**Hide This Folder/Hide This File** - removes the folder or file from the tree</br>
**Only Show This Folder** - remove all other folders and subfolders from the tree</br>
**Only Show This Folder And Subfolders** - remove other folders from the tree, but keep subfolders</br>
**Reset Folder Filter** - reset any folders previously filtered using the above</br>
**Toggle Badges** - enable/disable SCM decorations</br>
**Toggle Compact Folders** - enable/disable compressed folder names</br>
**Toggle Item Counts** - enable/disable counts of tags</br>
**Scan Open Files Only** - show TODOs from files open in VSCode (no search)</br>
**Scan Current File Only** - show TODOs from the current open file only</br>
**Scan Workspace And Open Files** - show TODOs from the workspace and any open files</br>
**Scan Workspace Only** - show TODOs from the workspace only (requires manual refresh)</br>
**Expand Tree/Collapse Tree** - expand or collapse the whole tree
**Show Tree View/Show Flat View/Show Tags Only View** - change the tree view style
**Group by Tag/Ungroup by Tag** - toggle grouping of items by tag
**Group by Sub Tag/Ungroup by Sub Tag** - toggle grouping of items by sub tag
**Export Tree** - dump the tree contents into a file
**Reveal Current File in Tree** - show the current editor file in tree (if present)

<sup>*Note: The current filters are shown in the debug log. Also, the filter can always be reset by right clicking the **Nothing Found** item in the tree. If your tree becomes invisible because everything is filtered and `hideTreeWhenEmpty` is set to true, you can reset the filter by pressing **F1** and selecting the **Todo Tree: Reset Folder Filter** command.*</sup>

## Commands

### Tags

To make it easier to configure the tags, there are two commands available:

**Todo Tree: Add Tag** - allows entry of a new tag for searching

**Todo Tree: Remove Tag** - shows a list of current tags which can be selected for removing

<sup>*Note: The Remove Tag command can be used to show current tags - just press Escape or Enter with out selecting any to close it.*</sup>

### Export

The contents of the tree can be exported using **Todo Tree: Export Tree**. A read-only file will be created using the path specified with `todo-tree.general.exportPath`. The file can be saved using **File: Save As...**. *Note: Currently **File: Save** does not work which seems to be a VSCode bug (see <https://github.com/microsoft/vscode/issues/101952>).*

### Switch Scope

**Todo Tree: Switch Scope** - shows a list of configured scopes which can be selected

### Navigation

The commands **Todo Tree: Go To Next** and **Todo Tree: Go To Previous** can be used to move to the next and previous TODO within the current editor.

## Configuration

The extension can be customised as follows (default values in brackets):

**todo-tree.general.debug** (`false`)</br>
Show a debug channel in the output view.

**todo-tree.general.periodicRefreshInteval** (`0`)</br>
Interval (in minutes) for automatically refreshing the tree. Set to '0' to disable, or to the number of minutes between refreshes. *Note: This is not typically needed as the tree will be refreshed when files change.*

**todo-tree.general.automaticGitRefreshInterval** (`0`)</br>
Polling interval (in seconds) for automatically refreshing the tree when your repository is updated. This will check if your repository HEAD has changed and trigger a rescan of the workspace if it has. This replaces the file watcher functionality. Set to '0' to disable, or to the number of seconds between checks.

**todo-tree.general.exportPath** (`~/todo-tree-%Y%m%d-%H%M.txt`)</br>
Path to use when exporting the tree. Environment variables will be expanded, e.g `${HOME}` and the path is passed through strftime (see <https://github.com/samsonjs/strftime>). Set the extension to `.json` to export as a JSON record.

**todo-tree.general.rootFolder** (`""`)</br>
By default, any open workspaces will have a tree in the view. Use this to force another folder to be the root of the tree. You can include environment variables and also use ${workspaceFolder}. e.g.</br>
`"todo-tree.general.rootFolder": "${workspaceFolder}/test"`</br>
or</br>
`"todo-tree.general.rootFolder": "${HOME}/project"`.</br>


<sup>*Note: Other open files (outside of the rootFolder) will be shown (as they are opened) with their full path in brackets.*</sup>

**todo-tree.general.schemes** (`['file','ssh','untitled']`)</br>
Editor schemes to find TODOs in. To find TODOs in settings files, for instance, add `vscode-userdata` or for output windows, add `output`.

**todo-tree.general.tags** (`["TODO","FIXME","BUG"]`)</br>
This defines the tags which are recognised as TODOs. This list is automatically inserted into the regex.

**todo-tree.general.tagGroups** (`{}`)</br>
This setting allows multiple tags to be treated as a single group. Example:

```json
    "todo-tree.general.tagGroups": {
        "FIXME": [
            "FIXME",
            "FIXIT",
            "FIX",
        ]
    },
```

This treats any of `FIXME`, `FIXIT` or `FIX` as `FIXME`. When the tree is grouped by tag, all of these will appear under the `FIXME` node. This also means that custom highlights are applied to the group, not each tag type.

<sup>*Note: all tags in the group should also appear in `todo-tree.general.tags`.*</sup>

**todo-tree.general.revealBehaviour** (`start of todo`)</br>
Change the cursor behaviour when double-clicking a todo in the tree. You can choose from: `start of todo` (moves the cursor to the beginning of the todo), `end of todo` (moves the cursor to the end of the todo) or `start of line` (moves the cursor to the start of the line).

**todo-tree.general.statusBar** (`none`)</br>
What to show in the status bar - nothing (`none`), total count (`total`), counts per tag (`tags`), counts for the top three tags (`top three`) or counts for the current file only (`current file`).

**todo-tree.general.statusBarClickBehaviour** (`reveal`)</br>
Set the behaviour of clicking the status bar to either cycle through the status bar display formats (`cycle`), reveal the tree (`reveal`) or to toggle highlights (`toggle highlights`).

**todo-tree.general.showIconsInsteadOfTagsInStatusBar** (`false`)</br>
Show icons instead of tags in the status bar.

**todo-tree.general.showActivityBarBadge** (`false`)</br>
Show a badge in the activity bar indicating the total number of found TODOs.

<sup>*Note: When the tree view is in the Explorer pane, the badge is displayed on the Explorer icon, which may not be desirable.*</sup>

**todo-tree.filtering.includeGlobs** (`[]`)</br>
Globs for use in limiting search results by inclusion, e.g. `[\"**/unit-tests/*.js\"]` to only show .js files in unit-tests subfolders. [Globs help](https://code.visualstudio.com/api/references/vscode-api#GlobPattern).

<sup>*Note: globs paths are absolute - not relative to the current workspace.*</sup>

**todo-tree.filtering.excludeGlobs** (`["**/node_modules/*/**"]`)</br>
Globs for use in limiting search results by exclusion (applied after **includeGlobs**), e.g. `[\"**/*.txt\"]` to ignore all .txt files.

<sup>*Note: `node_modules` are excluded by default.*</sup>

**todo-tree.filtering.includedWorkspaces** (`[]`)</br>
A list of workspace names to include as roots in the tree (wildcards can be used). An empty array includes all workspace folders.

**todo-tree.filtering.excludedWorkspaces** (`[]`)</br>
A list of workspace names to exclude as roots in the tree (wildcards can be used).

**todo-tree.filtering.passGlobsToRipgrep** (`true`)</br>
Set this to false to apply the globs *after* the search (legacy behaviour).

**todo-tree.filtering.useBuiltInExcludes** (`none`)</br>
Set this to use VSCode's built in files or search excludes. Can be one of `none`, `file excludes` (uses Files:Exclude), `search excludes` (Uses Search:Exclude) or `file and search excludes` (uses both).

**todo-tree.filtering.ignoreGitSubmodules** (`false`)</br>
If true, any subfolders containing a `.git` file will be ignored when searching.

**todo-tree.filtering.includeHiddenFiles** (`false`)</br>
If true, files starting with a period (.) will be included.

**todo-tree.highlights.enabled** (`true`)</br>
Set this to false to turn off highlighting.

**todo-tree.highlights.highlightDelay** (`500`)</br>
The delay before highlighting (milliseconds).

**todo-tree.highlights.defaultHighlight** (`{}`)</br>
Set default highlights. Example:

```json
{
    "foreground": "white",
    "background": "red",
    "icon": "check",
    "type": "text"
}
```

**todo-tree.highlights.customHighlight** (`{}`)</br>
Set highlights per tag (or tag group). Example:

```json
{
    "TODO": {
        "foreground": "white",

CHANGELOG.md
# Todo Tree Change Log

## v0.0.224 - 2023-02-09

- Remove filewatcher functionality
- Add a simple mechanism to watch for git repo changes instead of using the file watcher
- Add a simple periodic refresh

## v0.0.223 - 2023-02-01

- Flash the line containing the TODO when clicking on items in the tree view
- Fix go to next/previous TODO commands when regex matches the starts of the line
- Deprecate file watcher settings and remove from the README.md

## v0.0.222 - 2023-01-23

- Add commands to move to next and previous TODO in a file

## v0.0.221 - 2023-01-22

- Add default octicons for custom highlights (thanks to [Andrew Hlynskyi](https://github.com/ahlinc))
- Always show total in the activity bar regardless of other settings
- Add support for hiding counts from the activity bar badge
- Fix auto refresh when closing file (thanks to [yw662](https://github.com/yw662))
- Fix scopes documentation (thanks to [kjo-sdds](https://github.com/kjo-sdds))
- Update node_modules exclude to match VSCode (thanks to [Benas Svipas](https://github.com/svipas))

## v0.0.220 - 2022-11-07

- Fix missing 'new' keywords (thanks to [Elias](https://github.com/Elias-Graf))
- Fix duplicate entries when refreshing and files have unsaved changes (thanks to [Chris Bain](https://github.com/baincd))
- Fix issue with duplicate items (thanks to [Matt Toigo](https://github.com/mtoigo))
- Update octicons and fix todo tree icons (thanks to [Morgan](https://github.com/morgansliman))

## v0.0.219 - 2022-10-14

- Add activity bar badge support, but default to off and only for versions 172+ of vscode

## v0.0.218 - 2022-10-14

- Back out activity bar badge option for older versions of VSCode

## v0.0.217 - 2022-10-13

- Update minimum VSCode version to 1.172

## v0.0.216 - 2022-10-13

- Add support for activity bar badge

## v0.0.215 - 2022-01-04

- Update default regex in README.md
- Add "none" as a valid value for highlight type
- Add default icons for default tags
- Add markdown tags to default tag list
- Update screenshot in README.md
- Look for new location of VSCode's packaged ripgrep when the default setting is used

## v0.0.214 - 2021-09-07

- Build codicon list automatically
- Support highlighting from capture groups in the regex
- Fix default regex to not greedily match newlines
- Update default regex
- Fix group by tag
- When grouping by tag, don't add group node for single item
- Ensure markdown update popup is hidden for current session too
- Use entire URI for editor ID when highlighting
- Only show reveal button when not in tags only view
- use subtag match correctly
- Only show total count in title if tree counts are enabled

## v0.0.213 - 2021-04-27

- Remove some old migration code (before version 168)
- Fix colour validation
- Improve status bar layout
- Fix missing URI in multi line TODOs
- Add icon validation

## v0.0.212 - 2021-04-26

- Use the window icon for workspaces in the tree
- Move 'schemes' setting from highlights to general (and include "ssh" by default)
- Use URIs instead of file paths to improve support for remote file systems
- Fix group by tag in flat mode
- Show icons on status bar for each tag count
- Use tag order in status bar, unless showing top three
- Use tag order when grouping by tag

## v0.0.211 - 2021-04-09

- Fix file opening on Windows

## v0.0.210 - 2021-04-08

- Indicate current scope selection when changing
- Make reveal behaviour more consistent with VSCode

## v0.0.209 - 2021-03-28

- Add support for scopes, allowing predefined sets of globs to be selected, (thanks to [kkckkc](https://github.com/kkckkc))
- Add support for manually enabling multiline TODO support
- Improve multi-line todo handling
- Fix broken file icons in tree
- Fix missing filename when using compact folders

## v0.0.208 - 2021-03-20

- Revert commit which removed trailing text from multiline TODOs

## v0.0.207 - 2021-03-19

- Improve colour validation for icons
- Fix theme colours for ruler colour
- Remove trailing block comments when adding multiline todo to tree
- Support multi-line haskell comments
- Don't include trailing text to the end of multiline TODOs

## v0.0.206 - 2021-03-17

- Sort tags only view on tags order by default
- Ignore case when validating HTML colour names
- Allow highlights to be enabled/disabled by clicking the status bar
- Add ruler opacity setting
- Improve view title

## v0.0.205 - 2021-03-13

- Add support for VSCode issue reporter
- Fix highlights in split editors

## v0.0.204 - 2021-03-11

- Add improved markdown support dialog and settings update
- Fix expanded node storage after resetting cache

## v0.0.203 - 2021-03-10

- Fix highlighting of markdown TODOs
- Fix reversed sort order

## v0.0.202 - 2021-03-08

- Improve tag extraction when regex doesn't contain $TAGS
- Fix sorting (again)

## v0.0.201 - 2021-03-04

- Fix 'afterOfBefore' substitution

## v0.0.200 - 2021-03-04

- Add VS Code theme color reference link (thanks to [tksh164](https://github.com/tksh164))
- Add support for label (and tooltip) transformations (thanks to [TomFreudenberg](https://github.com/TomFreudenberg))
- Sort tags only view be label, not tag

## v0.0.199 - 2021-02-28

- Use storageUri instead of deprecated storagePath (v2)
- Add `**/node_modules` to default exclude globs
- Add support for removing trailing comment marker from .vue files (treated as HTML)
- Add context menu details to README.md

## v0.0.198 - 2021-02-25

- revert commit which replaced deprecated storage path

## v0.0.197 - 2021-02-24

- Add filter icon to filters active info node
- Fix missing context flag for show export button
- Implement a sub tag regex (based on work by[calw20](https://github.com/calw20))
- Support opening a URL by clicking on the sub tag in the tree
- Wait for startup to finish before starting extension
- Re-enable hideTreeWhenEmpty option
- Add delay before hiding tree when empty to prevent view disappearing on refresh
- Allow use of alternative "iconColor" in highlights configuration

src/icons.js
var vscode = require( 'vscode' );
var path = require( "path" );
var fs = require( 'fs' );
var octicons = require( '@primer/octicons' );

var utils = require( './utils.js' );
var attributes = require( './attributes.js' );
var themeColourNames = require( './themeColourNames.js' );
var codiconNames = require( './codiconNames.js' );
var productIconNames = require( './productIconNames.js' );

function getIcon( context, tag, debug )
{
    var colour = attributes.getIconColour( tag );

    var darkIconPath = context.asAbsolutePath( path.join( "resources/icons", "dark", "todo-green.svg" ) );
    var lightIconPath = context.asAbsolutePath( path.join( "resources/icons", "light", "todo-green.svg" ) );

    var compactColourName = colour;
    compactColourName = compactColourName.replace( /[^0-9a-zA-Z]/g, '' );

    var iconName = attributes.getIcon( tag );

    if( context.globalStorageUri.fsPath && !fs.existsSync( context.globalStorageUri.fsPath ) )
    {
        if( debug )
        {
            debug( "Attempting to create global storage folder " + context.globalStorageUri.fsPath );
        }
        fs.mkdirSync( context.globalStorageUri.fsPath );
    }

    if( iconName === 'todo-tree' || iconName === 'todo-tree-filled' )
    {
        if( utils.isThemeColour( colour ) )
        {
            colour = "green";
            compactColourName = "green";
        }

        var colouredTodoTreeIconPath = path.join( context.globalStorageUri.fsPath, iconName + "-" + compactColourName + ".svg" );
        if( !fs.existsSync( colouredTodoTreeIconPath ) )
        {
            var colouredTodoTreeIconDefinition = ( iconName === 'todo-tree' ) ?
                ( "<svg width=\"24\" height=\"24\" viewBox=\"0 0 24 24\" fill=\"none\" xmlns=\"http://www.w3.org/2000/svg\">" +
                    "<path d=\"M10.5268 1.22702C9.9895 1.4678 9.2269 2.20732 8.91492 2.79206C8.75894 3.08443 8.58562 3.13603 7.61503 3.08443C6.28048 3.01564 5.50054 3.30801 4.72061 4.18512C4.02733 4.97624 3.76736 6.00813 3.99267 7.00563L4.18332 7.77955L3.35139 8.50188C1.27156 10.3077 1.65287 13.541 4.07933 14.6589C4.56462 14.8996 5.10191 15.0888 5.27523 15.0888C5.43122 15.0888 5.70853 15.364 5.89918 15.6907C6.48846 16.671 8.30831 17.1354 9.48688 16.6023C10.0068 16.3615 10.1455 16.4991 10.4401 17.4794C10.6654 18.3221 10.2321 20.3859 9.50421 21.8305L8.91492 23H12.0347C14.4611 23 15 23 15 23L14.41 21.8305C13.68 20.39 13.6292 19.6979 13.4732 18.4253C13.3865 17.5998 13.4212 17.2386 13.6292 16.8258C13.9065 16.2927 13.9238 16.2927 14.4265 16.5507C15.605 17.1526 17.4769 16.7054 18.1008 15.6907C18.2915 15.364 18.5688 15.0888 18.7248 15.0888C18.8981 15.0888 19.4354 14.8996 19.9207 14.6589C22.3471 13.541 22.7284 10.3077 20.6486 8.50188L19.8167 7.77955L20.0073 7.00563C20.2326 6.00813 19.9727 4.97624 19.2794 4.18512C18.4995 3.30801 17.7195 3.01564 16.4023 3.08443C15.3277 3.13603 15.2757 3.11883 14.9638 2.62008C14.5478 1.93215 13.9238 1.39901 13.3346 1.17543C12.6586 0.917456 11.1507 0.951852 10.5268 1.22702Z\" stroke=\"" + colour + "\" stroke-width=\"1.5\" stroke-linejoin=\"bevel\" />" +
                    "<path d=\"M7 9.5L10.5 13L16.5 7\" stroke=\"" + colour + "\" stroke-width=\"1.5\" /></svg>" ) :
                ( "<?xml version=\"1.0\" standalone=\"no\"?><!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 20010904//EN\" \"http://www.w3.org/TR/2001/REC-SVG-20010904/DTD/svg10.dtd\">" +
                    "<svg version=\"1.0\" xmlns=\"http://www.w3.org/2000/svg\" width=\"28\" height=\"28\" viewBox=\"0 0 128.000000 128.000000\" preserveAspectRatio=\"xMidYMid meet\">" +
                    "<g transform=\"translate(0.000000,128.000000) scale(0.100000,-0.100000)\" fill=\"" + colour + "\" stroke=\"none\">" +
                    "<path d=\"M555 1266 c-31 -14 -75 -57 -93 -91 -9 -17 -19 -20 -75 -17 -77 4 -122 -13 -167 -64 -40 -46 -55 -106 -42 -164 l11 -45 -48 -42 c-120 -105 -98 -293 42 -358 28 -14 59 -25 69 -25 9 0 25 -16 36 -35 34 -57 139 -84 207 -53 30 14 38 6 55 -51 13 -49 -12 -169 -54 -253 l-34 -68 180 0 c140 0 178 3 172 12 -4 7 -15 24 -24 38 -24 34 -56 142 -65 216 -5 48 -3 69 9 93 16 31 17 31 46 16 68 -35 176 -9 212 50 11 19 27 35 36 35 10 0 41 11 69 25 140 65 162 " +
                    "253 42 358 l-48 42 11 45 c13 58 -2 118 -42 164 -45 51 -90 68 -166 64 -62 -3 -65 -2 -83 27 -24 40 -60 71 -94 84 -39 15 -126 13 -162 -3z m330 -236 c4 -6 -25 -45 -65 -86 -84 -89 -152 -190 -200 -297 -19 -42 -37 -77 -40 -77 -3 0 -18 23 -34 52 -16 28 -53 79 -82 114 -45 53 -51 64 -38 74 11 10 24 9 62 -5 26 -10 53 -22 60 -28 9 -7 42 21 125 106 133 137 193 178 212 147z\"/> </g></svg>" );

            fs.writeFileSync( colouredTodoTreeIconPath, colouredTodoTreeIconDefinition );
        }

        darkIconPath = colouredTodoTreeIconPath;
        lightIconPath = colouredTodoTreeIconPath;
    }
    else if( iconName && utils.isCodicon( iconName ) )
    {
        var themeColour;
        if( themeColourNames.indexOf( colour ) > -1 )
        {
            themeColour = new vscode.ThemeColor( colour );
        }
        return new vscode.ThemeIcon( iconName.trim().substr( 2, iconName.trim().length - 3 ), themeColour );
    }
    else if( iconName )
    {
        if( !octicons[ iconName ] )
        {
            iconName = "check";
        }

        if( utils.isThemeColour( colour ) )
        {
            colour = "green";
            compactColourName = "green";
        }

        if( context.globalStorageUri.fsPath )
        {
            var octiconIconPath = path.join( context.globalStorageUri.fsPath, "todo-" + iconName + "-" + compactColourName + ".svg" );
            if( !fs.existsSync( octiconIconPath ) )
            {
                var octiconIconDefinition = "<?xml version=\"1.0\" encoding=\"iso-8859-1\"?>\n" +
                    octicons[ iconName ].toSVG( { "xmlns": "http://www.w3.org/2000/svg", "fill": colour } );

                fs.writeFileSync( octiconIconPath, octiconIconDefinition );
            }

            darkIconPath = octiconIconPath;
            lightIconPath = octiconIconPath;
        }
    }
    else if( utils.isHexColour( colour ) || utils.isRgbColour( colour ) || utils.isNamedColour( colour ) )
    {
        var colouredIconPath = path.join( context.globalStorageUri.fsPath, "todo-" + compactColourName + ".svg" );
        if( !fs.existsSync( colouredIconPath ) )
        {
            var colouredIconDefinition =
                "<svg width=\"16\" height=\"16\" viewBox=\"0 0 16 16\" fill=\"none\" xmlns=\"http://www.w3.org/2000/svg\">" +
                "<path d=\"M11.1924 5.06738L11.8076 5.68262L6.6875 10.8027L4.19238 8.30762L4.80762 7.69238L6.6875 9.57227L11.1924 5.06738ZM8 1C8.64258 1 9.26237 1.08431 9.85938 1.25293C10.4564 1.41699 11.0146 1.65169 11.5342 1.95703C12.0537 2.25781 12.5254 2.6224 12.9492 3.05078C13.3776 3.47461 13.7422 3.94629 14.043 4.46582C14.3483 4.98535 14.583 5.54362 14.7471 6.14062C14.9157 6.73763 15 7.35742 15 8C15 8.64258 14.9157 9.26237 14.7471 9.85938C14.583 10.4564 14.3483 11.0146 14.043 11.5342C13.7422 12.0537 13.3776 12.5277 12.9492 12.9561C12.5254 13.3799 12.0537 13.7445 11.5342 14.0498C11.0146 14.3506 10.4564 14.5853 9.85938 14.7539C9.26237 14.918 8.64258 15 8 15C7.35742 15 6.73763 14.918 6.14062 14.7539C5.54362 14.5853 4.98535 14.3506 4.46582 14.0498C3.94629 13.7445 3.47233 13.3799 3.04395 12.9561C2.62012 12.5277 2.25553 12.0537 1.9502 11.5342C1.64941 11.0146 1.41471 10.4587 1.24609 9.86621C1.08203 9.26921 1 8.64714 1 8C1 7.35742 1.08203 6.73763 1.24609 6.14062C1.41471 5.54362 1.64941 4.98535 1.9502 4.46582C2.25553 3.94629 2.62012 3.47461 3.04395 3.05078C3.47233 2.6224 3.94629 2.25781 4.46582 1.95703C4.98535 1.65169 5.54134 1.41699 6.13379 1.25293C6.73079 1.08431 7.35286 1 8 1ZM8 14.125C8.56055 14.125 9.10059 14.0521 9.62012 13.9062C10.1442 13.7604 10.6318 13.5553 11.083 13.291C11.5387 13.0221 11.9535 12.7008 12.3271 12.3271C12.7008 11.9535 13.0199 11.541 13.2842 11.0898C13.5531 10.6341 13.7604 10.1465 13.9062 9.62695C14.0521 9.10742 14.125 8.5651 14.125 8C14.125 7.43945 14.0521 6.89941 13.9062 6.37988C13.7604 5.85579 13.5531 5.36816 13.2842 4.91699C13.0199 4.46126 12.7008 4.04655 12.3271 3.67285C11.9535 3.29915 11.5387 2.98014 11.083 2.71582C10.6318 2.44694 10.1442 2.23958 9.62012 2.09375C9.10059 1.94792 8.56055 1.875 8 1.875C7.43945 1.875 6.89714 1.94792 6.37305 2.09375C5.85352 2.23958 5.36589 2.44694 4.91016 2.71582C4.45898 2.98014 4.04655 3.29915 3.67285 3.67285C3.29915 4.04655 2.97786 4.46126 2.70898 4.91699C2.44466 5.36816 2.23958 5.85579 2.09375 6.37988C1.94792 6.89941 1.875 7.43945 1.875 8C1.875 8.56055 1.94792 9.10286 2.09375 9.62695C2.23958 10.1465 2.44466 10.6341 2.70898 11.0898C2.97786 11.541 3.29915 11.9535 3.67285 12.3271C4.04655 12.7008 4.45898 13.0221 4.91016 13.291C5.36589 13.5553 5.85352 13.7604 6.37305 13.9062C6.89258 14.0521 7.4349 14.125 8 14.125Z\"" +
                " fill=\"" + colour + "\" /></svg>";

            fs.writeFileSync( colouredIconPath, colouredIconDefinition );
        }

        darkIconPath = colouredIconPath;
        lightIconPath = colouredIconPath;
    }

    var icon = {
        dark: darkIconPath,
        light: lightIconPath
    };

    return icon;
}

function validateIcons( workspace )
{
    function checkIcon( setting )
    {
        var icon = workspace.getConfiguration( 'todo-tree.highlights' ).get( setting + ".icon" );
        if( icon !== undefined )
        {
            if( utils.isCodicon( icon ) )
            {
                var codicon = icon.substr( 2, icon.length - 3 );
                if( codiconNames.indexOf( codicon ) === -1 && productIconNames.indexOf( codicon ) === -1 )
                {
                    invalidIcons.push( setting + '.icon(' + icon + ')' );
                }
            }
            else if( !octicons[ icon ] && ( icon !== 'todo-tree' || icon !== 'todo-tree-filled' ) )
            {
                invalidIcons.push( setting + '.icon(' + icon + ')' );
            }
        }
    }

    var invalidIcons = [];
    var result = "";

    checkIcon( 'defaultHighlight' );

    var config = vscode.workspace.getConfiguration( 'todo-tree.highlights' );
    Object.keys( config.customHighlight ).forEach( function( tag )
    {
        checkIcon( 'customHighlight.' + tag );
    } );

    if( invalidIcons.length > 0 )
    {
        result = "Invalid icons: " + invalidIcons.join( ', ' ) + ".";
    }

    return result;
}

module.exports.getIcon = getIcon;
module.exports.validateIcons = validateIcons;
